### ISSUE TITLE
Review-2018.07.15-吴晓丽

### ISSUE THREAD

**Issue by ghost** [DIYer22/weekly-review#202]
1.李宏毅视频课看到了反向传播
2.课程的Sumamry未完成
------------------------------------Summary-----------------------------

### 1. Introduction of this course

- 深度学习包含于机器学习，机器学习包含于人工智能。
人工智能包括人类设定好的天生本能和机器学习。
人工智能是我们要达成的目标，机器学习是我们实现人工智能的手段。
人工智能不是依靠人类设定来实现的！人工智能不是写程序让机器做事，而是写程序让机器具有学习的能力。

- 机器学习≈让机器具有一种能力——根据你提供的资料，寻找出我们要寻找的function

- 如何寻找function？
寻找function分为以下三个步骤：
第一：找一个function set（又叫Model）
第二：让机器具备衡量一个function好坏的能力(Loss function)
第三：让机器有一个自动的方法（好的演算法），能从function set中挑一个最好的function

- 
![1](https://user-images.githubusercontent.com/28921270/42734986-4e9e2aec-887f-11e8-934d-252f329ff2d3.png)

1. Scenario：学习的情境，根据手头上的data选择scenario，通常scenario是没有办法自己控制的

2. Task：要解的问题，根据要找的function的output的形式选择task

3. Method：求解的方法，又叫Model 

- Regression：
要机器输出的结果是个scalar

- Classification：
Binary Classification相当于判断题，输出是Y/N
Multi-class Classification相当于选择题，输出是Class1/
Class2/···/Class N

- scenario的选择：
A 理想状况：拥有大量的labelled data，选择Supervised Learning
B拥有少量的labelled data+大量unlabeled data，选择Semi-supervised Learning
C 拥有少量的labelled data+大量的与我们现在考虑的问题没什么关系的data，选择Transfer Learning
//D Unsupervised Learning
//E Structured Learning
//F Reinforcement Learning

### 2. Regression
（1）预测宝可梦进化后的CP值
![2](https://user-images.githubusercontent.com/28921270/42735045-7b3c0c76-8880-11e8-9770-c31216296a02.png)

- 第一步：选定一个function set
![3](https://user-images.githubusercontent.com/28921270/42735057-a26ab446-8880-11e8-962d-8d5e43c8285e.png)

- 第二步：让机器可以判断一个function的好坏
用Loss function来衡量function set中某个function的好坏
Loss function 的 output 越小，function越好
![4](https://user-images.githubusercontent.com/28921270/42735072-d2054b30-8880-11e8-8e2e-ec3ce6a93770.png)
![5](https://user-images.githubusercontent.com/28921270/42735155-dc96a264-8881-11e8-812f-c0d8f258b8a3.png)

- 第三步：找一个使得Loss function最小的function
  这里使用了Gradient Descent
  这里η是指learning rate（学习率/步长）
![6](https://user-images.githubusercontent.com/28921270/42735156-dce84614-8881-11e8-9f3d-edd8d9f96844.png)
Loss function分别对参数w和b求微分，初始值为w0、b0，以学习率η不断地对w和b进行更新，寻找极值点（微分为0的点）
将最终的w和b带入Model，便得到最好的function
（2）Generalization

- 选定以下5个线性模型，分别进行Gradient Descent求得各个模型下的best function
![7](https://user-images.githubusercontent.com/28921270/42735157-dd2bb91c-8881-11e8-8ff9-216c8cd5f2d6.png)
得到以下结论：
①选择的Model次数越高，Training data的Average Error就越小，但是我们真正关注的是testing data的average error。
②然而，在Testing Data上，越复杂的Model不一定总是会有更好的表现。这就是Overfitting。

- 如何解决Overfitting？
①收集更多的收据，看看是否存在之前Model未考虑的隐藏因素，重新对Model进行设计
②修改Loss function，加入Regularization term。加入Regularization term的loss不仅可以帮助我们找到最好的function并且最好的function是相对平滑的

### 3. Where does the error come from?
1.error的来源：bias（f*的平均值与f target的距离）和variance（用同一Model、不同data训练得到的f*们是否集中）
2.对一个相同的Model，使用100组data分别进行训练，由于training data的不同，我们会得到不同的best function，记作f*。
![8](https://user-images.githubusercontent.com/28921270/42735158-dd6e129e-8881-11e8-9ab4-b13ad01872d2.png)

- Model越复杂， variance越大；Model越简单， variance越小

- Model越复杂， bias越小；Model越简单，bias越大

- 当Error来自variance很大时，叫做overfitting

- 当Error来自bias很大时，叫做underfitting

3.Diagnosis
李宏毅：硕士论文如果用到Machine Learning，最后要思考是bias大还是variance大？并分析！

- Performance：如果你的Model不能很好地拟合所有的训练点时
Conclusion：你的bias大
Solution：重新设计你的Model：
可以通过增加更多的features;
可以选择一个更为复杂的model

- Performance：如果你的Model可以很好地拟合training data，    但是testing data的error却很大时
Conclusion：可能是你的variance大
Solution：
more data；优点：不会伤害bias 缺点：可能数据不好捕获
regularization，使loss function更平滑，但可能会影响bias（缺点）。需要trade-off between bias and variance。

4.我们手头上拥有的testing data（public）是有它自己的bias的，并不能代表真正的testing data（真正testing data是你没有的，private），真正的testing data的Average Error很可能比你手头上的testing set的Average error大！
我们不应该：简单地把Training Set全都用来训练，然后从中选出最小Error的Model了事！
我们应该：Cross Validation 或者N-fold Cross Validation

- Cross Validation（交叉验证）
![9](https://user-images.githubusercontent.com/28921270/42735160-ddbb671a-8881-11e8-8093-e9ddc24d2346.png)
1.	将原本的Training Set分成Training Set和Validation Set两部分。
2.	用被分出来的Training Set部分进行Model训练
3.	用被分出来的Validation set部分进行Model验证；观察不同的Model在Validation Set上的表现：
①你可以直接选出Error最小的Model3
②你也可以定好Error最小的Model3，然后用原本的大Training Set再次进行训练，选出Model3下的best function

- N-fold Cross Validation（N重交叉验证）
![10](https://user-images.githubusercontent.com/28921270/42735161-de62f34a-8881-11e8-90f6-7f2894ffda46.png)
1.	将Training Set分成三份，每次选一份做Validation Set，剩下的两份作Training Set，那么就有C31=3种不同的分配方式
2.	计算每个Model的不同分配方式下的Error，然后计算每个Model的Average Error
3.	选定Average Error最小的Model1
4.	然后对Model1用原本的大Training Set再次进行训练，选出Model1中的best function

未完待续·······
:octocat:
